1.	Import the housing data as a data frame and ensure that the data is loaded properly.

In [3]:
import pandas as pd

#import the CSV
df = pd.read_csv("C:\\Users\\hessk\\OneDrive\\Desktop\\DSC550\\train.csv")

#display the first few rows to ensure it loaded properly
print(df.head(5))
print(df.info(5))

   Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD   

2.	Drop the "Id" column and any features that are missing more than 40% of their values.

In [13]:
#drop the Id column
df = df.drop(columns=["Id"])

print(df.head)

<bound method NDFrame.head of       MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0             60       RL         65.0     8450   Pave   NaN      Reg   
1             20       RL         80.0     9600   Pave   NaN      Reg   
2             60       RL         68.0    11250   Pave   NaN      IR1   
3             70       RL         60.0     9550   Pave   NaN      IR1   
4             60       RL         84.0    14260   Pave   NaN      IR1   
...          ...      ...          ...      ...    ...   ...      ...   
1455          60       RL         62.0     7917   Pave   NaN      Reg   
1456          20       RL         85.0    13175   Pave   NaN      Reg   
1457          70       RL         66.0     9042   Pave   NaN      Reg   
1458          20       RL         68.0     9717   Pave   NaN      Reg   
1459          20       RL         75.0     9937   Pave   NaN      Reg   

     LandContour Utilities LotConfig  ... PoolArea PoolQC  Fence MiscFeature  \
0            

In [15]:
#drop the columns missing more than 40%, which means keeping the data that has more than 60%. 
#setting a threshold of 60% allows me to drop any that do not meet that threshold.
threshold = len(df) * 0.6  
df = df.dropna(axis=1, thresh=threshold)

3.	For numerical columns, fill in any missing data with the median value.

In [20]:
#identify the numeric columns
num_cols = df.select_dtypes(include=['number']).columns

print(num_cols)

Index(['MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond',
       'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2',
       'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF',
       'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath',
       'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces',
       'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF',
       'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'MiscVal',
       'MoSold', 'YrSold', 'SalePrice'],
      dtype='object')


In [5]:
#find the median for each numeric column and display them
medians = df.median(numeric_only=True)

print("Median values:")
print(medians)

Median values:
Id                  730.5
MSSubClass           50.0
LotFrontage          69.0
LotArea            9478.5
OverallQual           6.0
OverallCond           5.0
YearBuilt          1973.0
YearRemodAdd       1994.0
MasVnrArea            0.0
BsmtFinSF1          383.5
BsmtFinSF2            0.0
BsmtUnfSF           477.5
TotalBsmtSF         991.5
1stFlrSF           1087.0
2ndFlrSF              0.0
LowQualFinSF          0.0
GrLivArea          1464.0
BsmtFullBath          0.0
BsmtHalfBath          0.0
FullBath              2.0
HalfBath              0.0
BedroomAbvGr          3.0
KitchenAbvGr          1.0
TotRmsAbvGrd          6.0
Fireplaces            1.0
GarageYrBlt        1980.0
GarageCars            2.0
GarageArea          480.0
WoodDeckSF            0.0
OpenPorchSF          25.0
EnclosedPorch         0.0
3SsnPorch             0.0
ScreenPorch           0.0
PoolArea              0.0
MiscVal               0.0
MoSold                6.0
YrSold             2008.0
SalePrice        163000

In [22]:
#take the missing data, and fill in the na with the median of the columns. 
df[num_cols] = df[num_cols].fillna(df[num_cols].median())

4.	For categorical columns, fill in any missing data with the most common value (mode).

In [9]:
#identify the categorical columns and display them
cat_cols = df.select_dtypes(include=['object', 'category']).columns

print(cat_cols)

Index(['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities',
       'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2',
       'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st',
       'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation',
       'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
       'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual',
       'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual',
       'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature',
       'SaleType', 'SaleCondition'],
      dtype='object')


In [13]:
#find and display the mode of the categorical columns
modes = df[cat_cols].mode(dropna=True).iloc[0]

print(modes)

MSZoning              RL
Street              Pave
Alley               Grvl
LotShape             Reg
LandContour          Lvl
Utilities         AllPub
LotConfig         Inside
LandSlope            Gtl
Neighborhood       NAmes
Condition1          Norm
Condition2          Norm
BldgType            1Fam
HouseStyle        1Story
RoofStyle          Gable
RoofMatl         CompShg
Exterior1st      VinylSd
Exterior2nd      VinylSd
MasVnrType       BrkFace
ExterQual             TA
ExterCond             TA
Foundation         PConc
BsmtQual              TA
BsmtCond              TA
BsmtExposure          No
BsmtFinType1         Unf
BsmtFinType2         Unf
Heating             GasA
HeatingQC             Ex
CentralAir             Y
Electrical         SBrkr
KitchenQual           TA
Functional           Typ
FireplaceQu           Gd
GarageType        Attchd
GarageFinish         Unf
GarageQual            TA
GarageCond            TA
PavedDrive             Y
PoolQC                Gd
Fence              MnPrv


In [19]:
#fill in only the missing values with the mode. 
#using a loop lets the data from the correct columns to fill in appropriately. 
cat_cols = df.select_dtypes(include=['object', 'category']).columns
for col in cat_cols:
    if df[col].isnull().any():
        mode_val = df[col].mode(dropna=True)
        if not mode_val.empty:
            df[col] = df[col].fillna(mode_val[0])


5.	Convert the categorical columns to dummy variables.

In [22]:
#convert the categorical columns into dummy columns
df = pd.get_dummies(df, columns=cat_cols, drop_first=True)

print(df.head(5))

   Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0   1          60         65.0     8450            7            5       2003   
1   2          20         80.0     9600            6            8       1976   
2   3          60         68.0    11250            7            5       2001   
3   4          70         60.0     9550            7            5       1915   
4   5          60         84.0    14260            8            5       2000   

   YearRemodAdd  MasVnrArea  BsmtFinSF1  ...  SaleType_ConLI  SaleType_ConLw  \
0          2003       196.0         706  ...           False           False   
1          1976         0.0         978  ...           False           False   
2          2002       162.0         486  ...           False           False   
3          1970         0.0         216  ...           False           False   
4          2000       350.0         655  ...           False           False   

   SaleType_New  SaleType_Oth  SaleTyp

6.	Split the data into a training and test set, where the SalePrice column is the target.

In [42]:
from sklearn.model_selection import train_test_split

#set the target and feature
y = df['SalePrice']
X = df.drop(columns=['SalePrice'])

In [44]:
#split the data into an 80/20 split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

7.	Run a linear regression and report the R2-value and RMSE on the test set.

In [47]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error
import numpy as np

#fit the linear regression
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [49]:
#predict using the test set
y_pred = model.predict(X_test)


In [53]:
#calculate R2 and RMSE
r2 = r2_score(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))

print("R2-value test:", r2)
print("RMSE test:", rmse)

R2-value test: 0.6478367401194003
RMSE test: 51973.138076035146


8.	Fit and transform the training features with a PCA so that 90% of the variance is retained (see section 9.1 in the Machine Learning with Python Cookbook).

In [55]:
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import joblib

#standardize the PCA
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)


In [57]:
#fit PCA with 90% variance
pca = PCA(n_components=0.90, svd_solver='full')
X_train_pca = pca.fit_transform(X_train_scaled)

In [65]:
#save the transformed features
pd.DataFrame(X_train_pca).to_csv('X_train_pca.csv', index=False)
joblib.dump(pca, 'pca_model.joblib')
joblib.dump(scaler, 'scaler.joblib')

['scaler.joblib']

9.	How many features are in the PCA-transformed matrix?

In [59]:
#print the shape which gives the the number of features
print(X_train_pca.shape[1])

127


10.	Transform but DO NOT fit the test features with the same PCA.

In [67]:
#use the fitted traning
scaler = joblib.load('scaler.joblib')
pca = joblib.load('pca_model.joblib')

In [69]:
#standardize the featues on the test data
X_test_scaled = scaler.transform(X_test)

In [71]:
#transform the test data
X_test_pca = pca.transform(X_test_scaled)

11.	Repeat step 7 with your PCA transformed data.

In [73]:
#fit the regression
model = LinearRegression()
model.fit(X_train_pca, y_train)

LinearRegression()

In [75]:
#predict using the test PCA set
y_pred = model.predict(X_test_pca)


In [77]:
#calculate R2 and RMSE
r2 = r2_score(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))

print("R2-value PCA:", r2)
print("RMSE PCA:", rmse)

R2-value on test set (PCA features): 0.8408543593955071
RMSE on test set (PCA features): 34938.502799751244


12.	Take your original training features (from step 6) and apply a min-max scaler to them.

In [81]:
#min-max scale training and test set
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

13.	Find the min-max scaled features in your training set that have a variance above 0.1 (see Section 10.1 in the Machine Learning with Python Cookbook).

In [99]:
#calculate the variances
variances = X_train_scaled.var()

In [101]:
#find features above 0.1
high_variance_features = variances[variances > 0.1]

print("Features with variance above 0.1:")
print(high_variance_features)

Features with variance above 0.1:
[0.12828308]


14.	Transform but DO NOT fit the test features with the same steps applied in steps 11 and 12.

In [105]:
#split into training and test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [107]:
#min-max on training
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)

In [109]:
#transform without fitting
X_test_scaled = scaler.transform(X_test)

In [111]:
#fit PCA to training
pca = PCA(n_components=0.9, svd_solver='full')
X_train_pca = pca.fit_transform(X_train_scaled)


In [113]:
#transform test, without fitting
X_test_pca = pca.transform(X_test_scaled)

15.	Repeat step 7 with the high variance data.

In [125]:
#find R2 and RMSE
r2 = r2_score(y_test, y_pred)
rmse = root_mean_squared_error(y_test, y_pred)

print(f"R² value high variance: {r2:.4f}")
print(f"RMSE value high variance: {rmse:.2f}")

R² value high variance: -10570565081393809850368.0000
RMSE value high variance: 9004425710093457.00


16.	Summarize your findings.

High Variance: having an extremely large R2 that is negative means my model is not capturing any meaningful relationships. This is not a good model to use. A high RMSE value means the predictions are not accurate at predicting anything. 

PCA: The R2 for the PCA was .84 meaning 84% of the time the variance was accounted for. The R2 says it is a good fit for the data, could be improved, but is a good fit.  RMSE of 34,938 is not terrible, there are numerous variables in numerous sizes, it could use some modificiation, but it is a decent place. It is also the best fit of the 3.

Initial model: The R2 of .65 meaning 65% of the time the variance was accounted for. This is better than flipping a coin, but not by a whole lot. There is a lot that oculd be modified with this yet to train it better. RMSE of 51,973 is worse than the PCA. There is much modification that could be done on this model. 

